### PR TITLE
refactor(cli): gluegun-inspired improvements

### PR DESCRIPTION
**Pull request by galvez**
As per recent discussion in #core -- this makes options specific to commands and adopts gluegun's standard for commands as modules. This is a stepping stone to more changes in the new CLI project.

### CODE REVIEW

File: packages/cli/test/unit/build.test.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ describe('build', () => {
 
   beforeAll(async () => {
     build = await import('../../src/commands/build')
-    build = build.default
+    build = build.default.run
```
Comment (pi0) [inline packages/cli/test/unit/build.test.js]: import().then(m => m.default.run). One less line :D

File: packages/cli/src/options/index.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import common from './common'
+import network from './network'
+
+export default {
+  ...common,
+  ...network
```
[the remark this answers is not in the document]
Comment (galvez) [reply packages/cli/src/options/index.js]: @pi0 yep, makes sense, I'll change.

File: packages/cli/src/run.js [hunk 3]
```diff
@@ -26,7 +37,7 @@ export default function run() {
   })
 
   return commands[cmd]() // eslint-disable-line import/namespace
-    .then(m => m.default())
+    .then(prepareAndRun)
```
[the remark this answers is not in the document]
Comment (galvez) [reply packages/cli/src/run.js]: @pi0 gotcha, I'll modify.

File: packages/cli/src/run.js [hunk 4]
```diff
@@ -1,7 +1,18 @@
...
 import setup from './setup'
 
+const prepareAndRun = async (commandModule) => {
+  commandModule = commandModule.default
+  const nuxtCmd = new NuxtCommand({
+    usage: commandModule.usage,
+    description: commandModule.description,
+    options: commandModule.options
+  })
+  await commandModule.run(nuxtCmd)
```
[the remark this answers is not in the document]
Comment (galvez) [reply packages/cli/src/run.js]: @pi0 ha, yeah, was unsure about this. Thanks.

File: packages/cli/src/commands/build.js [hunk 5]
```diff
@@ -1,46 +1,76 @@
...
-    usage: 'build <dir>',
-    options: [ 'analyze', 'quiet' ]
-  })
-
-  const argv = nuxtCmd.getArgv()
+export default {
+  description: 'Compiles the application for production deployment',
+  usage: 'build <dir>',
+  options: {
+    ...commonOptions,
```
[the remark this answers is not in the document]
Comment (galvez) [reply packages/cli/src/commands/build.js]: @pi0 fixed :)

**Comment by pi0** [comment]
gym-sick-moves-12X2n39NZ2PoZO

File: packages/cli/src/options/common.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+export default {
```
[the remark this answers is not in the document]
Comment (galvez) [reply packages/cli/src/options/common.js]: @pi0 common options are added on demand. A module-defined command may simply not define them. If they're not defined, they don't show up in the help message :)

[the remark this answers is not in the document]
Comment (pi0) [reply packages/cli/src/options/common.js]: True. But thare are at least 3 essential configs:

- `-h, `--help`: Help message is auto-generated. So it makes sense that every command at least supports it.
- `-v`, `--version`: Nuxt / CLI version?
- `-c`, `--config`: Finally if we want to introduce custom options directly inside `nuxt.config.js` or auto-registration of CLI commands from modules included in `modules: []` section of config and also for correctly determining `rootDir`, `appDir` (and other customized configs) we need a way that users can specify that config file.

BTW the last item I've mentioned is not probably implemented ATM but during CLI refactor we can see the requirements.

**Comment by pimlie** [comment]
Looks good to me, the small remarks I could have had where already addressed by @pi0 :heart: 

Personally I would've preferred more of a OOP approach so that each command export is fully self-contained / self-supported (so more like oclif). Especially looking at those tests I just dont like that `wrapAndRun` method, to an outsider that probably seems just a random method.

This could also be a possible implementation:
```js
// command export in ./src/commands/*
export default new NuxtCommand({
  options: ...,
  description: ...,
  run() { // <- dont use (arrow) function so you can just use this. within run to access NuxtCommand stuff
  }
})
// and in ./src/run.js
.then(command => {
  if (command.constructor.name === 'Object') {
    command = new NuxtCommand(command)
  }
  return command.run()
}).catch...

```

**Comment by galvez** [comment]
@pimlie I thought about that too, but I didn't want to turn importing NuxtCommand into a requirement for module command writers. Keeping a simple object signature seems more developer-friendly to me.

**Comment by pimlie** [comment]
@galvez In my example it isn't a requirement either, see the change to `src/index` (which could probably be a createNuxtCommand method to replace wrapAndRun). I dont think we should forcibly wrap into NuxtCommand when its just not necessary. What e.g. if I want to extend NuxtCommand to add some scaffolding for my own commands?
Also I think we should try to avoid to both wrap and run in the same method, better to keep that separated. My example above gives the developer a choice, either export a standard js object with some props or export a full (derived) NuxtCommand instance with whatever you wanted to change.

**Comment by galvez** [comment]
@pimlie I can see your points, but I think there's benefit in always providing a NuxtCommand instance, so we offer the maximum access methods to developers. It'll also play a role in harmonizing future module commands, but I'll leave those thoughts to @pi0 now :)

**Review by pi0 — APPROVED** [review]

**Pull request merged**